<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ simDNAmixtures</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">simDNAmixtures</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/getting_started.html">Getting started</a></li>
    <li><a class="dropdown-item" href="../articles/simulating_YSTR_profiles.html">Simulating YSTR profiles</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="SMASH_to_wide_table.html">SMASH_to_wide_table()</a></code>

        </dt>
        <dd>Converts SMASH (SampleName, Marker, Allele, Size, Height) data to a wide table</dd>
      </dl><dl><dt>

          <code><a href="allele_specific_stutter_model.html">allele_specific_stutter_model()</a></code>

        </dt>
        <dd>Stutter model where the expected stutter rate depends on the allele and locus</dd>
      </dl><dl><dt>

          <code><a href="drop_model.html">drop_model()</a></code>

        </dt>
        <dd>Defines a (semi-continuous) drop model</dd>
      </dl><dl><dt>

          <code><a href="gamma_model.html">gamma_model()</a></code>

        </dt>
        <dd>Defines a gamma model for peak height variability</dd>
      </dl><dl><dt>

          <code><a href="gf_configuration.html">gf_configuration()</a></code>

        </dt>
        <dd>Stutters and size regressions for a GlobalFiler 3500 kit</dd>
      </dl><dl><dt>

          <code><a href="global_stutter_model.html">global_stutter_model()</a></code>

        </dt>
        <dd>Global stutter model where the expected stutter rate is constant across alleles and loci</dd>
      </dl><dl><dt>

          <code><a href="kits.html">kits</a></code>

        </dt>
        <dd>Properties such as loci, dye, sizes for most standard forensic kits</dd>
      </dl><dl><dt>

          <code><a href="log_normal_model.html">log_normal_model()</a></code>

        </dt>
        <dd>Defines a log normal model for peak height variability</dd>
      </dl><dl><dt>

          <code><a href="read_STRmix_kit_settings.html">read_STRmix_kit_settings()</a></code>

        </dt>
        <dd>Read STRmix Kit Settings</dd>
      </dl><dl><dt>

          <code><a href="read_allele_freqs.html">read_allele_freqs()</a></code>

        </dt>
        <dd>Read allele frequencies in FSIgen format (.csv)</dd>
      </dl><dl><dt>

          <code><a href="read_size_regression.html">read_size_regression()</a></code>

        </dt>
        <dd>Reads a size regression file</dd>
      </dl><dl><dt>

          <code><a href="read_stutter_exceptions.html">read_stutter_exceptions()</a></code>

        </dt>
        <dd>Reads a stutter exceptions file with overrides for expected stutter ratios</dd>
      </dl><dl><dt>

          <code><a href="read_stutter_regression.html">read_stutter_regression()</a></code>

        </dt>
        <dd>Reads a stutter regression file</dd>
      </dl><dl><dt>

          <code><a href="read_wide_table.html">read_wide_table()</a></code>

        </dt>
        <dd>Read wide table (.txt) with Allele1, Allele2, ... columns as is</dd>
      </dl><dl><dt>

          <code><a href="sample_LSAE.html">sample_LSAE()</a></code>

        </dt>
        <dd>Sample Locus Specific Amplification Efficiency (LSAE) according to prior</dd>
      </dl><dl><dt>

          <code><a href="sample_contributor_genotypes.html">sample_contributor_genotypes()</a></code>

        </dt>
        <dd>Sample genotypes for mixture contributors according to allele frequencies</dd>
      </dl><dl><dt>

          <code><a href="sample_drop_model.html">sample_drop_model()</a></code>

        </dt>
        <dd>Sample drop model(s) with parameters according to priors</dd>
      </dl><dl><dt>

          <code><a href="sample_gamma_model.html">sample_gamma_model()</a></code>

        </dt>
        <dd>Sample gamma model(s) with parameters according to priors</dd>
      </dl><dl><dt>

          <code><a href="sample_genotype.html">sample_genotype()</a></code>

        </dt>
        <dd>Sample a genotype according to allele frequencies</dd>
      </dl><dl><dt>

          <code><a href="sample_log_normal_model.html">sample_log_normal_model()</a></code>

        </dt>
        <dd>Sample log normal model(s) with parameters according to priors</dd>
      </dl><dl><dt>

          <code><a href="sample_log_normal_parameters.html">sample_log_normal_parameters()</a></code>

        </dt>
        <dd>Sample log normal model parameters according to priors</dd>
      </dl><dl><dt>

          <code><a href="sample_log_normal_stutter_variance.html">sample_log_normal_stutter_variance()</a></code>

        </dt>
        <dd>Sample log normal stutter variance parameters according to priors</dd>
      </dl><dl><dt>

          <code><a href="sample_many_pedigree_genotypes.html">sample_many_pedigree_genotypes()</a></code>

        </dt>
        <dd>Sample (many) genotypes for pedigree members according to allele frequencies by allele dropping and possibly taking linkage into account by simulating recombination.</dd>
      </dl><dl><dt>

          <code><a href="sample_mixture_from_genotypes.html">sample_mixture_from_genotypes()</a></code>

        </dt>
        <dd>Sample mixture profile with provided genotypes</dd>
      </dl><dl><dt>

          <code><a href="sample_mixtures.html">sample_mixtures()</a></code>

        </dt>
        <dd>Sample mixtures with random genotypes and random parameters according to priors</dd>
      </dl><dl><dt>

          <code><a href="sample_mixtures_fixed_parameters.html">sample_mixtures_fixed_parameters()</a></code>

        </dt>
        <dd>Sample mixtures with known genotypes and fixed parameters</dd>
      </dl><dl><dt>

          <code><a href="sample_mixtures_from_genotypes.html">sample_mixtures_from_genotypes()</a></code>

        </dt>
        <dd>Sample mixtures with known genotypes and random parameters according to priors</dd>
      </dl><dl><dt>

          <code><a href="sample_offspring.html">sample_offspring()</a></code>

        </dt>
        <dd>Sample offspring from two parental genotypes</dd>
      </dl><dl><dt>

          <code><a href="sample_pedigree_genotypes.html">sample_pedigree_genotypes()</a></code>

        </dt>
        <dd>Sample genotypes for pedigree according to allele frequencies by allele dropping.</dd>
      </dl><dl><dt>

          <code><a href="stutter_type.html">stutter_type()</a></code>

        </dt>
        <dd>Defines a stutter type to be used in the allele specific stutter model.</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maarten Kruijver.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

