<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting started • simDNAmixtures</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">simDNAmixtures</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/getting_started.html">Getting started</a></li>
    <li><a class="dropdown-item" href="../articles/simulating_YSTR_profiles.html">Simulating YSTR profiles</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Getting started</h1>
                        <h4 data-toc-skip class="author">Maarten
Kruijver</h4>
            
            <h4 data-toc-skip class="date">2025-04-14</h4>
      

      <div class="d-none name"><code>getting_started.Rmd</code></div>
    </div>

    
    
<p>Simulation studies are often used in forensic genetics to study the
behaviour of probabilistic genotyping software. In many cases,
researchers use simplified models that ignore complexities such as peak
height variability and stutters. The goal of this package is to make it
straightforward to use probabilistic genotyping models in simulation
studies.</p>
<p>For a general introduction to forensic genetics and probabilistic
genotyping, see <span class="citation">(Buckleton, Bright, and Taylor
2018)</span>. See also <span class="citation">(Gill et al. 2021)</span>
for a recent review of probabilistic genotyping systems.</p>
<div class="section level2">
<h2 id="basic-log-normal-example">Basic log normal example<a class="anchor" aria-label="anchor" href="#basic-log-normal-example"></a>
</h2>
<p>Below, we load allele frequencies provided with the package and
demonstrate how mixtures can be sampled using a log normal model for
peak heights (refer to <span class="citation">(Taylor, Bright, and
Buckleton 2013)</span> for details on the model). The template
parameters for each contributor are picked uniformly between 50 and
10,000. The degradation parameter is picked from a gamma distribution
with shape 2.5 and scale 0.001.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">freqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_allele_freqs.html">read_allele_freqs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"FBI_extended_Cauc_022024.csv"</span>,</span>
<span>                                       package <span class="op">=</span> <span class="st">"simDNAmixtures"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gf_configuration.html">gf_configuration</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sampling_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>min_template <span class="op">=</span> <span class="fl">50.</span>, max_template <span class="op">=</span> <span class="fl">10000.</span>,</span>
<span>                            degradation_shape <span class="op">=</span> <span class="fl">2.5</span>, degradation_scale <span class="op">=</span> <span class="fl">1e-3</span><span class="op">)</span></span></code></pre></div>
<p>After the setup, a single function call is sufficient to generate a
set of samples. We set <code>n=2</code> to generate two samples. The
contributors to the sample are named <code>U1</code> and <code>U2</code>
which means that each mixed sample has two unrelated contributors.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">mixtures</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sample_mixtures.html">sample_mixtures</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2</span>, contributors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"U1"</span>, <span class="st">"U2"</span><span class="op">)</span>, freqs <span class="op">=</span> <span class="va">freqs</span>,</span>
<span>                            sampling_parameters <span class="op">=</span> <span class="va">sampling_parameters</span>,</span>
<span>                            model_settings <span class="op">=</span> <span class="va">gf</span><span class="op">$</span><span class="va">log_normal_bwfw_settings</span>,</span>
<span>                            sample_model <span class="op">=</span> <span class="va">sample_log_normal_model</span><span class="op">)</span></span></code></pre></div>
<p>The <code>mixtures</code> object contains the simulation results. The
<code>sample_mixtures</code> function has an optional
<code>results_directory</code> argument. If the path to a directory is
provided, then all simulation results will be persisted on disk in file
formats that are readily imported in probabilistic genotyping software.
We inspect the <code>parameter_summary</code> property of the result of
<code>sample_mixtures</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">mixtures</span><span class="op">$</span><span class="va">parameter_summary</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="33%">
<col width="16%">
<col width="16%">
<col width="20%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="left">SampleName</th>
<th align="left">contributors1</th>
<th align="left">contributors2</th>
<th align="left">model</th>
<th align="right">template1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">sample_0001_N2_7753_6977_ca</td>
<td align="left">U1</td>
<td align="left">U2</td>
<td align="left">log_normal_model</td>
<td align="right">7752.690</td>
</tr>
<tr class="even">
<td align="left">sample_0002_N2_5356_7818_bb</td>
<td align="left">U1</td>
<td align="left">U2</td>
<td align="left">log_normal_model</td>
<td align="right">5356.177</td>
</tr>
</tbody>
</table>
<p>The first sample is in approximately a 2:1 ratio. We print the first
10 peaks of the profile below.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mixtures</span><span class="op">$</span><span class="va">samples</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">mixture</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Locus</th>
<th align="left">Allele</th>
<th align="right">Height</th>
<th align="right">Size</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">D3S1358</td>
<td align="left">14</td>
<td align="right">1481</td>
<td align="right">117.33</td>
</tr>
<tr class="even">
<td align="left">D3S1358</td>
<td align="left">15</td>
<td align="right">22260</td>
<td align="right">121.40</td>
</tr>
<tr class="odd">
<td align="left">D3S1358</td>
<td align="left">16</td>
<td align="right">6660</td>
<td align="right">125.48</td>
</tr>
<tr class="even">
<td align="left">vWA</td>
<td align="left">13</td>
<td align="right">174</td>
<td align="right">164.80</td>
</tr>
<tr class="odd">
<td align="left">vWA</td>
<td align="left">14</td>
<td align="right">4054</td>
<td align="right">168.84</td>
</tr>
<tr class="even">
<td align="left">vWA</td>
<td align="left">17</td>
<td align="right">588</td>
<td align="right">180.95</td>
</tr>
<tr class="odd">
<td align="left">vWA</td>
<td align="left">18</td>
<td align="right">7710</td>
<td align="right">184.99</td>
</tr>
<tr class="even">
<td align="left">vWA</td>
<td align="left">19</td>
<td align="right">4561</td>
<td align="right">189.02</td>
</tr>
<tr class="odd">
<td align="left">D16S539</td>
<td align="left">8</td>
<td align="right">200</td>
<td align="right">239.58</td>
</tr>
<tr class="even">
<td align="left">D16S539</td>
<td align="left">9</td>
<td align="right">5827</td>
<td align="right">243.61</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="basic-gamma-example">Basic gamma example<a class="anchor" aria-label="anchor" href="#basic-gamma-example"></a>
</h2>
<p>We repeat the example above with a gamma distribution for peak
heights instead of a log normal one. The mean peak height parameter
<code>mu</code> is chosen uniformly between 50 and 5,000. The variance
parameter <code>cv</code> parameter
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>σ</mi><annotation encoding="application/x-tex">\sigma</annotation></semantics></math>
in <span class="citation">(Bleka, Storvik, and Gill 2016)</span>) is
chosen uniformly between 0.05 and 0.35. The degradation parameter is
sampled from a Beta distribution with parameters 10 and 1, which means
most profile will be only mildly degraded.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gamma_sampling_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>min_mu <span class="op">=</span> <span class="fl">50.</span>, max_mu <span class="op">=</span> <span class="fl">5e3</span>,</span>
<span>                            min_cv <span class="op">=</span> <span class="fl">0.05</span>, max_cv <span class="op">=</span> <span class="fl">0.35</span>,</span>
<span>                            degradation_shape1 <span class="op">=</span> <span class="fl">10</span>, degradation_shape2 <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>We now invoke the <code>sample_mixtures</code> function again to
sample two mixtures; both consisting of two unrelated contributors. We
do not sample stutters.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">mixtures</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sample_mixtures.html">sample_mixtures</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2</span>, contributors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"U1"</span>, <span class="st">"U2"</span><span class="op">)</span>, freqs <span class="op">=</span> <span class="va">freqs</span>,</span>
<span>                            sampling_parameters <span class="op">=</span> <span class="va">gamma_sampling_parameters</span>,</span>
<span>                            model_settings <span class="op">=</span> <span class="va">gf</span><span class="op">$</span><span class="va">gamma_settings_no_stutter</span>,</span>
<span>                            sample_model <span class="op">=</span> <span class="va">sample_gamma_model</span><span class="op">)</span></span></code></pre></div>
<p>We see that the first sample has
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>μ</mi><annotation encoding="application/x-tex">\mu</annotation></semantics></math>
around 1,236 and is in a 65:35 ratio.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">mixtures</span><span class="op">$</span><span class="va">parameter_summary</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span> </span></code></pre></div>
<table class="table">
<colgroup>
<col width="42%">
<col width="20%">
<col width="20%">
<col width="17%">
</colgroup>
<thead><tr class="header">
<th align="left">SampleName</th>
<th align="left">contributors1</th>
<th align="left">contributors2</th>
<th align="left">model</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">sample_0001_N2_4297_48_52_bb</td>
<td align="left">U1</td>
<td align="left">U2</td>
<td align="left">gamma_model</td>
</tr>
<tr class="even">
<td align="left">sample_0002_N2_4513_34_66_aa</td>
<td align="left">U1</td>
<td align="left">U2</td>
<td align="left">gamma_model</td>
</tr>
</tbody>
</table>
<p>The coefficient of variation for a full heterozygote is about 9% for
the first sample.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">mixtures</span><span class="op">$</span><span class="va">parameter_summary</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">5</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="36%">
<col width="26%">
<col width="26%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="left">SampleName</th>
<th align="right">mixture_proportions1</th>
<th align="right">mixture_proportions2</th>
<th align="right">mu</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">sample_0001_N2_4297_48_52_bb</td>
<td align="right">0.4822929</td>
<td align="right">0.5177071</td>
<td align="right">4297.329</td>
</tr>
<tr class="even">
<td align="left">sample_0002_N2_4513_34_66_aa</td>
<td align="right">0.3395792</td>
<td align="right">0.6604208</td>
<td align="right">4513.154</td>
</tr>
</tbody>
</table>
<p>We print the first 10 peaks of the profile below.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mixtures</span><span class="op">$</span><span class="va">samples</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">mixture</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Locus</th>
<th align="left">Allele</th>
<th align="right">Height</th>
<th align="right">Size</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">D3S1358</td>
<td align="left">15</td>
<td align="right">3604</td>
<td align="right">121.40</td>
</tr>
<tr class="even">
<td align="left">D3S1358</td>
<td align="left">17</td>
<td align="right">2812</td>
<td align="right">129.56</td>
</tr>
<tr class="odd">
<td align="left">D3S1358</td>
<td align="left">18</td>
<td align="right">1978</td>
<td align="right">133.64</td>
</tr>
<tr class="even">
<td align="left">vWA</td>
<td align="left">14</td>
<td align="right">2359</td>
<td align="right">168.84</td>
</tr>
<tr class="odd">
<td align="left">vWA</td>
<td align="left">16</td>
<td align="right">4001</td>
<td align="right">176.91</td>
</tr>
<tr class="even">
<td align="left">vWA</td>
<td align="left">18</td>
<td align="right">2151</td>
<td align="right">184.99</td>
</tr>
<tr class="odd">
<td align="left">D16S539</td>
<td align="left">11</td>
<td align="right">3291</td>
<td align="right">251.67</td>
</tr>
<tr class="even">
<td align="left">D16S539</td>
<td align="left">12</td>
<td align="right">2749</td>
<td align="right">255.70</td>
</tr>
<tr class="odd">
<td align="left">CSF1PO</td>
<td align="left">10</td>
<td align="right">1303</td>
<td align="right">298.34</td>
</tr>
<tr class="even">
<td align="left">CSF1PO</td>
<td align="left">11</td>
<td align="right">2973</td>
<td align="right">302.30</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-bleka2016euroformix" class="csl-entry">
Bleka, Øyvind, Geir Storvik, and Peter Gill. 2016. <span>“EuroForMix: An
Open Source Software Based on a Continuous Model to Evaluate STR DNA
Profiles from a Mixture of Contributors with Artefacts.”</span>
<em>Forensic Science International: Genetics</em> 21: 35–44. <a href="https://doi.org/10.1016/j.fsigen.2015.11.008" class="external-link">https://doi.org/10.1016/j.fsigen.2015.11.008</a>.
</div>
<div id="ref-buckleton2018forensic" class="csl-entry">
Buckleton, John S, Jo-Anne Bright, and Duncan Taylor. 2018. <em>Forensic
DNA Evidence Interpretation</em>. CRC press.
</div>
<div id="ref-gill2021review" class="csl-entry">
Gill, Peter, Corina Benschop, John Buckleton, Øyvind Bleka, and Duncan
Taylor. 2021. <span>“A Review of Probabilistic Genotyping Systems:
EuroForMix, DNAStatistX and STRmix™.”</span> <em>Genes</em> 12 (10):
1559. <a href="https://doi.org/10.3390/genes12101559" class="external-link">https://doi.org/10.3390/genes12101559</a>.
</div>
<div id="ref-taylor2013interpretation" class="csl-entry">
Taylor, Duncan, Jo-Anne Bright, and John Buckleton. 2013. <span>“The
Interpretation of Single Source and Mixed DNA Profiles.”</span>
<em>Forensic Science International: Genetics</em> 7 (5): 516–28. <a href="https://doi.org/10.1016/j.fsigen.2013.05.011" class="external-link">https://doi.org/10.1016/j.fsigen.2013.05.011</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maarten Kruijver.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
